### ISSUE TITLE
(typescript) Improving the Typescript grammar

### ISSUE THREAD

**Issue by ChristianIvicevic** [highlightjs/highlight.js#2190]
TL;DR: I am writing a lot of Typescript lately and coming from Markdown to Asciidoctor (which uses highlight.js) I am really missing the contextual parsing and highlighting of variable names, function invocations and possibly types (at some point).

In this ticket I am showing you examples that made me notice that highlight.js isn't actually parsing code semantically. Furthermore in a second post in this ticket I am making some proposals for technical analysis.

---

For training material and a lot of code that I am writing, I am in the process of migrating over from Markdown to AsciiDoc. Missing some of the highlighting I had previously is really noticeable as you can see in this example:

<img width="538" alt="Screenshot 2019-10-12 at 19 28 42" src="https://user-images.githubusercontent.com/45569997/66705407-86588880-ed26-11e9-879a-eaf713756466.png">

<img width="567" alt="Screenshot 2019-10-12 at 19 29 06" src="https://user-images.githubusercontent.com/45569997/66705409-95d7d180-ed26-11e9-9f3f-1cf41efd824b.png">

<img width="597" alt="Screenshot 2019-10-12 at 19 49 53" src="https://user-images.githubusercontent.com/45569997/66705631-80b07200-ed29-11e9-9e23-240a7b471ff6.png">

<img width="508" alt="Screenshot 2019-10-12 at 19 50 00" src="https://user-images.githubusercontent.com/45569997/66705633-85752600-ed29-11e9-996e-dc5e47e70389.png">

The same code snippet is more difficult to read (which is highly subjective of course) and I personally appreciate the distinction of classes/types from object members and function invocations etc. which is helpful for beginners to distinguish various code elements.

_Originally posted by @ChristianIvicevic in https://github.com/highlightjs/highlight.js/issues/1378#issuecomment-541346850_

**Comment by ChristianIvicevic** [comment]
I do believe parsing 3rd party classes is one of the most difficult tasks when working with Typescript source files. This is due to their nature in possibly being intertwined with various syntactical elements and decorations.

Really naive example: assuming everything after the colon is a type you might run into a literal union `doSomething(arg: "div" | "span")` and all of a sudden you run into a scenario wherein you expected a type but got a literal value - are falling back to format that as strings or forcefully format that using your type CSS class?

Another example is with regards to the generics. As you can see, `T` and `K` are contextually picked up being types as well even though they are referred to on multiple occasions in different contexts.

My suggestion is to analyze first whether some more naive features can be implemented to improve highlighting and ignoring more complex cases in the beginning:

* Anything of the form `foo(...)` or `bar.foo(...)` that seems to be a function invocation can be highlighted (in my screenshot the `map` method)
* Parameters inside a function declaration, i.e. in the scope of Typescript anything of the form `name: Type` inside parens such as `entity`, `obj` or `propertyNames` without highlighting their next occurences again

Furthermore Typescript exposes its own compiler and parser that can be picked up to analyze the AST generated which might come in handy to dekegate parsing Typescript code to the actual compiler.

Depending on the parsing options you get the following output:

![image](https://user-images.githubusercontent.com/45569997/66714828-26142600-edbc-11e9-81cb-c51f2ca8a015.png)

Alternatively even the following:

![image](https://user-images.githubusercontent.com/45569997/66714838-375d3280-edbc-11e9-862e-57ca536b4c82.png)


Resources:

* https://basarat.gitbooks.io/typescript/docs/compiler/parser.html
* https://ts-ast-viewer.com
* https://github.com/dsherret/ts-morph
* https://github.com/TypeStrong/ntypescript

**Comment by yyyc514** [comment]
> Furthermore Typescript exposes its own compiler and parser that can be picked up to analyze the AST generated which might come in handy to dekegate parsing Typescript code to the actual compiler.

This is probably out-of-scope... at least until we split the parsing from the HTML generation stuff (which isn't a high priority), then PERHAPS someone could build an alternative parser that laid our HTML generation on top of it... right now there just isn't any infrastructure at all for something like that... you'd be talking another whole highlighting engine.

> Anything of the form foo(...) or bar.foo(...) that seems to be a function invocation can be highlighted (in my screenshot the map method)

That's doable (esp after my look-ahead patch).  I already have Javascript grammars of my own that do this, but I don't know if that will ever make it into Highlight.js proper.  The philosophy in the past has been a bit more minimal (we purposely don't highlight function calls, punctuation, operators, etc)...

The first form might be harder in that you have to know it's an invocation and not a definition.

> Parameters inside a function declaration, i.e. in the scope of Typescript anything of the form name: Type inside parens such as entity, obj or propertyNames without highlighting their next occurences again

Seems like that would be some basic pattern matching and improving the parsing of function declarations would be something that fits here.  So that'd be the type of PR we'd be happy to review I think.

**Comment by ChristianIvicevic** [comment]
After thinking about it, this can be difficult as well. Take my snippets from the other issues I created lately:

```typescript
[1, 2, 3].reduce((acc, item) => [...acc, item], [0]);
```

Even though `acc` and `item` have no type annotations (since they are infered) they are parameters and not arguments. Distinguishing this from function declarations will be more difficult than I expected.

Regarding your gripes about the philosophy: is it reasonable to consider some experimental or more sophisticated variants of languages and grammars? Say the current basic ones that only naively highlight keywords and built ins and a different one you have to explicitly opt in that has even more features. Just thinking out loud right now.

**Comment by yyyc514** [comment]
> Regarding your gripes about the philosophy: is it reasonable to consider some experimental or more sophisticated variants of languages and grammars?

It wasn't a gripe, it was more of a heads up.

Well, sure it's reasonable to play around and see what's possible (and I'm interested in those experiments also).  Also no one can stop you from releasing your own grammar that does who know what crazy stuff.  And people could choose to use it or not (I'm interesting in making that type of thing easier - for people willing to do a small amount of work to build their own library, etc).

But with core you have not only philosophy but also consistency.  A lot of people use the library and expect consistent results across languages (as much as is reasonable).  Failure to do so results in bug reports, issues, etc... so even if we all agreed this would be "really nice" the problem becomes by adding it to one language only you create 184 potential issues we have to deal with for all the other grammars... and that's a net loss.

And it seems few people (though correct me if I'm wrong) are interested in a project like this spanning ALL the grammars... everyone has their own language - and I can understand why for familiarity, etc)...

It's very hard as a contributor to keep track of all the different syntaxes and understand what is valid here and not there, etc. :-)

---

I think the best bet for something like this in general is to have flags (and there has been some discussion of that)... so that you could say `highlight: { functionInvokation: true }` and that would turn on support for that in some grammars, but it wouldn't be a default we'd ship enabled.

And that would allow us to slip in more features quietly without creating a huge expectation that those things will work for all languages.

**Comment by yyyc514** [comment]
I think the best way to push big things like that is become a serious contributor... champion Typescript (if that's your thing)... fix all the small issues... then when you start suggesting bigger issues and showing us experiments it'll be more natural to listen.

Right now the severe limiting factor is contributor time... lots of idea, so little time.

**Comment by yyyc514** [comment]
Closing this issue. See above discussion.  In general Highlight.js does not highlight variables, function invocations, or user types - so there is really nothing to be done here. We're analyzing the code with pattern matching, not semantically.

> Furthermore Typescript exposes its own compiler and parser that can be picked up to analyze the AST generated which might come in handy to dekegate parsing Typescript code to the actual compiler.

If someone wanted to work on a 3rd party plugin that used the TS parser and integrated with Highlight.js I'd be happy to discuss that. See discussion over at:

https://github.com/highlightjs/highlight.js/issues/2212

I'd guess that would greatly increase the download size (for web environments), although for server-side environments it might be more acceptable.
